### PR TITLE
Fixes #1790: Support MongoDB Extended JSON (v2) in apoc.mongodb.* procedures parameters

### LINKED ISSUE #1308: Create new Mongo Procedures
The new set of procedures (`apoc.mongo.*`) should leverage:
 * https://docs.mongodb.com/manual/reference/connection-string/
 * map configuration param

### LINKED ISSUE #1790: Support MongoDB Extended JSON (v2) in apoc.mongodb.* procedures parameters
## Feature description (Mandatory)
It would be very useful if the MongoDB Extended JSON (v2) would be supported in all apoc.mongodb.* procedures query parameter. I´m not too familiar with Mongo, but maybe it´s useful for the update parameter too.

Relates to issue #1784 

## Considered alternatives
No alternatives.

Current apoc.mongodb could be extended:
The simplest way would be to serialize the query map to json string and create a new bson document with Document.parse(). As [parse](https://mongodb.github.io/mongo-java-driver/4.2/apidocs/bson/org/bson/Document.html#parse(java.lang.String)) should already support the extended format. This surely may not be the optimized way to this.
If the simple implementation will be preferred a new config parameter could be introduced (useExtendedFormat) so existing use cases (usage without extended format) will not be interfered (referred to preformance issues)

## How this feature can improve the project?
The main concern would be to query by an ObjectId in all procedures. This is the default identification field for mongo documents.

### PR DESCRIPTION
**Pull request by vga91**
- Fixes #1308
Added procedures `apoc.mongo.aggregate/count/find/insert/update/delete`

- Fixes #1790 (for new procedues)
Added conditions in `compatibleValues` to support all possible extjson types

### CODE REVIEW

File: docs/asciidoc/modules/ROOT/pages/database-integration/mongodb.adoc [hunk 1]
```diff
@@ -1,7 +1,11 @@
 [[mongodb]]
-= MongoDB
+= MongoDB (Deprecated)
 :description: This section describes procedures that can be used to interact with MongoDB.
 
+[NOTE]
+====
+This procedures are deprecated in favor of xref::database-integration/mongo.adoc[apoc.mongo.* procedures] .
```
Comment (fbiville) [inline docs/asciidoc/modules/ROOT/pages/database-integration/mongodb.adoc:7]:
```suggestion
These procedures are deprecated in favor of xref::database-integration/mongo.adoc[apoc.mongo.* procedures] .
```

File: docs/asciidoc/modules/ROOT/pages/overview/apoc.ttl/index.adoc [hunk 2]
```diff
@@ -22,6 +22,6 @@ CALL apoc.ttl.expireIn(node,timeDelta,'time-unit') - expire node after specified
 
 
 |label:function[]
-|label:apoc-core[]
```
Comment (fbiville) [inline docs/asciidoc/modules/ROOT/pages/overview/apoc.ttl/index.adoc:25]: seems unrelated to the PR

File: docs/asciidoc/modules/ROOT/partials/generated-documentation/documentation.adoc [hunk 3]
```diff
@@ -196,7 +196,6 @@ apoc.bitwise.op(60,'\|',13) bitwise operations a & b, a \| b, a ^ b, ~a, a >> b,
 |label:apoc-core[]
 |===
 
-[discrete]
```
Comment (fbiville) [inline docs/asciidoc/modules/ROOT/partials/generated-documentation/documentation.adoc:199]: this change seems unrelated to the PR scope

File: docs/asciidoc/modules/ROOT/partials/generated-documentation/documentation.adoc [hunk 4]
```diff
@@ -212,6 +211,11 @@ apoc.bolt.execute(url-or-key, kernelTransaction, params, config) - access to oth
 apoc.bolt.load(url-or-key, kernelTransaction, params, config) - access to other databases via bolt for read
 |label:procedure[]
 |label:apoc-full[]
+|xref::overview/apoc.bolt/apoc.bolt.load.fromLocal.adoc[apoc.bolt.load.fromLocal icon:book[]]
```
Comment (fbiville) [inline docs/asciidoc/modules/ROOT/partials/generated-documentation/documentation.adoc:214]: this change seems unrelated to the PR scope

File: docs/asciidoc/modules/ROOT/partials/generated-documentation/documentation.adoc [hunk 5]
```diff
@@ -3096,7 +3136,7 @@ CALL apoc.ttl.expireIn(node,timeDelta,'time-unit') - expire node after specified
 
 
 |label:function[]
-|label:apoc-core[]
+|label:apoc-full[]
```
Comment (fbiville) [inline docs/asciidoc/modules/ROOT/partials/generated-documentation/documentation.adoc:3139]: this change seems unrelated to the PR scope

File: full/src/main/java/apoc/mongodb/Mongo.java [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+                getExceptionConsumer("apoc.mongo.aggregate", uri, config));
+    }
+
+    @Procedure("apoc.mongo.count")
+    @Description("apoc.mongo.count(uri, query, $config) yield value - perform a count operation on mongodb collection")
+    public Stream<LongResult> count(@Name("uri") String uri, @Name("query") Object query, @Name(value = "config", defaultValue = "{}") Map<String, Object> config) {
+        MongoDbConfig conf = new MongoDbConfig(config);
+        return executeMongoQuery(uri, conf, coll -> {
+            long count = coll.count(getDocument(query));
+            return Stream.of(new LongResult(count));
```
Comment (fbiville) [inline full/src/main/java/apoc/mongodb/Mongo.java:47]: could this just return a `LongResult`?

File: full/src/main/java/apoc/mongodb/Mongo.java [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+            mongoErrorLog("apoc.mongo.insert", uri, config, e, "documents = " + documents + ",");
+            throw new RuntimeException(e);
+        }
+    }
+
+    @Procedure("apoc.mongo.update")
+    @Description("apoc.mongo.update(uri, query, update, $config) - updates the given documents from the mongodb collection and returns the number of affected documents")
+    public Stream<LongResult> update(@Name("uri") String uri, @Name("query") Object query, @Name("update") Object update, @Name(value = "config", defaultValue = "{}") Map<String, Object> config) {
+        MongoDbConfig conf = new MongoDbConfig(config);
+        return executeMongoQuery(uri, conf, coll -> Stream.of(new LongResult(coll.update(getDocument(query), getDocument(update)))),
```
Comment (fbiville) [inline full/src/main/java/apoc/mongodb/Mongo.java:76]: same question as above

File: full/src/main/java/apoc/mongodb/Mongo.java [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+                getExceptionConsumer("apoc.mongo.update", uri, config, "query = " + query + ",  update = " + update + ","));
+    }
+
+    @Procedure("apoc.mongo.delete")
+    @Description("apoc.mongo.delete(uri, query, $config) - delete the given documents from the mongodb collection and returns the number of affected documents")
+    public Stream<LongResult> delete(@Name("uri") String uri, @Name("query") Object query, @Name(value = "config", defaultValue = "{}") Map<String, Object> config) {
+        MongoDbConfig conf = new MongoDbConfig(config);
+        return executeMongoQuery(uri, conf, coll -> Stream.of(new LongResult(coll.delete(getDocument(query)))),
```
Comment (fbiville) [inline full/src/main/java/apoc/mongodb/Mongo.java:84]: same

File: full/src/test/java/apoc/mongodb/MongoTestBase.java [hunk 9]
```diff
@@ -0,0 +1,273 @@
...
+                "price", 1200,
+                "tags", Arrays.asList("Tech", "Mobile", "Phone", "Android"))));
+        productReferences = StreamSupport.stream(productCollection.find().spliterator(), false)
+                .map(doc -> (ObjectId) doc.get("_id"))
+                .collect(Collectors.toList());
+        boughtListObjectIds = productReferences.stream().map(ObjectId::toString).collect(Collectors.toList());
+
+        personCollection.insertOne(new Document(map("name", "Andrea Santurbano",
+                "bought", productReferences,
+                "born", DateUtils.parseDate("11-10-1935", "dd-MM-yyyy"),
```
Comment (fbiville) [inline full/src/test/java/apoc/mongodb/MongoTestBase.java:150]: you look so young for your age @conker84 :D

File: full/src/test/java/apoc/mongodb/MongoTest.java [hunk 10]
```diff
@@ -0,0 +1,497 @@
...
+                    row -> {
+                    });
+        } catch (Exception e) {
+            assertTrue(e.getMessage().contains("not authorized on foo to execute command"));
+            throw e;
+        }
+    }
+
+    @Test
+    public void shouldNotFailIfUriHasNonRootUserNotAuthorizedInDb() {
```
Comment (fbiville) [inline full/src/test/java/apoc/mongodb/MongoTest.java:142]:
```suggestion
    public void shouldNotFailIfUriHasRootUser() {
```

**Review by fbiville — COMMENTED** [review]
we'll review the second part next week, since this is a big chunk of work

File: full/src/main/java/apoc/mongodb/Mongo.java (hunk 6 above)
Comment (vga91) [reply full/src/main/java/apoc/mongodb/Mongo.java:47]: As a procedure, I have to return a `Stream` because of `org.neo4j.internal.kernel.api.exceptions.ProcedureException: Procedures must return a Stream of records`. 
We could return a `long` by turning it into function, but for consistency with the old procedures, I think it's better this way.

File: full/src/main/java/apoc/mongodb/Mongo.java (hunk 7 above)
Comment (vga91) [reply full/src/main/java/apoc/mongodb/Mongo.java:76]: As above

File: full/src/main/java/apoc/mongodb/Mongo.java (hunk 8 above)
Comment (vga91) [reply full/src/main/java/apoc/mongodb/Mongo.java:84]: As above

File: full/src/main/java/apoc/mongodb/Mongo.java (hunk 6 above)
Comment (JMHReif) [reply full/src/main/java/apoc/mongodb/Mongo.java:47]: Ok. All good. Thanks for explaining!

File: full/src/main/java/apoc/mongodb/MongoDB.java [hunk 11]
```diff
@@ -164,75 +162,18 @@ public void insert(@Name("host") String hostOrKey, @Name("db") String db, @Name(
                     Map<String, Object> result = coll.first(Map.of(conf.getIdFieldName(), new ObjectId(objectIdValue)));
                     return result == null || result.isEmpty() ? Stream.empty() : Stream.of(new MapResult(result));
                 },
-                e -> log.error("apoc.mongo.get.byObjectId - hostOrKey = [" + hostOrKey + "], db = [" + db + "], collection = [" + collection + "], objectIdValue = [" + objectIdValue + "]",e));
+                e -> log.error("apoc.mongodb.get.byObjectId - hostOrKey = [" + hostOrKey + "], db = [" + db + "], collection = [" + collection + "], objectIdValue = [" + objectIdValue + "]",e));
```
Comment (JMHReif) [inline full/src/main/java/apoc/mongodb/MongoDB.java:165]: Why did this change? Shouldn't it be `apoc.mongo.update` (same name in error as proc we're calling)?

File: full/src/main/java/apoc/mongodb/MongoDBUtils.java [hunk 12]
```diff
@@ -0,0 +1,94 @@
...
+        }
+
+        class Factory {
+            public static Coll create(String url, String db, String coll, boolean compatibleValues,
+                                      boolean extractReferences,
+                                      boolean objectIdAsMap) {
+                try {
+                    return new MongoDBColl(url, db, coll, compatibleValues, extractReferences, objectIdAsMap);
+                } catch (Exception e) {
+                    throw new RuntimeException("Could not create MongoDBClientWrapper instance", e);
```
Comment (JMHReif) [inline full/src/main/java/apoc/mongodb/MongoDBUtils.java:49]: Should the MongoDBClientWrapper change? Didn't we change this name?

**Review by JMHReif — APPROVED** [review]

File: full/src/main/java/apoc/mongodb/MongoDB.java (hunk 11 above)
Comment (vga91) [reply full/src/main/java/apoc/mongodb/MongoDB.java:165]: This is a modification not strictly related to pr.
I noticed that the procedure is called `apoc.mongodb.get.byObjectId`,
so the error message here was wrong.

File: full/src/main/java/apoc/mongodb/MongoDBUtils.java (hunk 12 above)
Comment (vga91) [reply full/src/main/java/apoc/mongodb/MongoDBUtils.java]: Yes that's right, I hadn't noticed. Now I have changed it.

**Review by JMHReif — APPROVED** [review]
